File: slp/backend-dotnet/Features/Admin/AdminService.cs
using backend_dotnet.Features.Comment;
using backend_dotnet.Features.Helpers;
using backend_dotnet.Features.Quiz;
using backend_dotnet.Features.Session;
using backend_dotnet.Features.User;
using System.Text.Json;

namespace backend_dotnet.Features.Admin;

public class AdminService : IAdminService
{
    private readonly IUserRepository _userRepo;
    private readonly IQuizRepository _quizRepo;
    private readonly ICommentRepository _commentRepo;
    private readonly IAdminLogRepository _logRepo;
    private readonly ISessionRepository _sessionRepo;

    public AdminService(
        IUserRepository userRepo,
        IQuizRepository quizRepo,
        ICommentRepository commentRepo,
        IAdminLogRepository logRepo,
        ISessionRepository sessionRepo)
    {
        _userRepo = userRepo;
        _quizRepo = quizRepo;
        _commentRepo = commentRepo;
        _logRepo = logRepo;
        _sessionRepo = sessionRepo;
    }

    // --- Users ---
    public async Task<PaginatedResult<UserDto>> GetAllUsersAsync(string? search = null, int page = 1, int pageSize = 20)
    {
        var (users, total) = await _userRepo.GetAllAsync(search, page, pageSize);
        var dtos = users.Select(u => new UserDto
        {
            Id = u.Id,
            Username = u.Username,
            Email = u.Email,
            EmailConfirmed = u.EmailConfirmed,
            Role = u.Role,
            Status = u.Status,
            CreatedAt = u.CreatedAt
        }).ToList();

        return new PaginatedResult<UserDto>
        {
            Items = dtos,
            Total = total,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<bool> BanUserAsync(int adminId, int userId)
    {
        var user = await _userRepo.GetByIdAsync(userId);
        if (user == null || user.Role == "admin") return false;

        user.Status = "banned";
        await _userRepo.UpdateAsync(user);
        await _sessionRepo.RevokeAllForUserAsync(userId);

        await _logRepo.LogAsync(new AdminLog
        {
            AdminId = adminId,
            Action = "ban_user",
            TargetType = "user",
            TargetId = userId
        });
        return true;
    }

    public async Task<bool> UnbanUserAsync(int adminId, int userId)
    {
        var user = await _userRepo.GetByIdAsync(userId);
        if (user == null) return false;

        user.Status = "active";
        await _userRepo.UpdateAsync(user);

        await _logRepo.LogAsync(new AdminLog
        {
            AdminId = adminId,
            Action = "unban_user",
            TargetType = "user",
            TargetId = userId
        });
        return true;
    }

    // --- Quizzes ---
    public async Task<PaginatedResult<QuizAdminDto>> GetAllQuizzesAsync(string? search = null, int page = 1, int pageSize = 20)
    {
        var (quizzes, total) = await _quizRepo.GetAllForAdminAsync(search, page, pageSize);
        var dtos = quizzes.Select(q => new QuizAdminDto
        {
            Id = q.Id,
            Title = q.Title,
            UserId = q.UserId,
            Username = q.User?.Username ?? "deleted",
            Visibility = q.Visibility,
            Disabled = q.Disabled,
            CreatedAt = q.CreatedAt
        }).ToList();

        return new PaginatedResult<QuizAdminDto>
        {
            Items = dtos,
            Total = total,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<bool> DisableQuizAsync(int adminId, int quizId)
    {
        var quiz = await _quizRepo.GetByIdAsync(quizId, includeDisabled: true);
        if (quiz == null) return false;

        quiz.Disabled = true;
        await _quizRepo.UpdateAsync(quiz);

        await _logRepo.LogAsync(new AdminLog
        {
            AdminId = adminId,
            Action = "disable_quiz",
            TargetType = "quiz",
            TargetId = quizId
        });
        return true;
    }

    public async Task<bool> EnableQuizAsync(int adminId, int quizId)
    {
        var quiz = await _quizRepo.GetByIdAsync(quizId, includeDisabled: true);
        if (quiz == null) return false;

        quiz.Disabled = false;
        await _quizRepo.UpdateAsync(quiz);

        await _logRepo.LogAsync(new AdminLog
        {
            AdminId = adminId,
            Action = "enable_quiz",
            TargetType = "quiz",
            TargetId = quizId
        });
        return true;
    }

    // --- Comments ---
    public async Task<PaginatedResult<CommentAdminDto>> GetAllCommentsAsync(bool includeDeleted = false, string? search = null, int page = 1, int pageSize = 20)
    {
        var (comments, total) = await _commentRepo.GetAllAsync(includeDeleted, search, page, pageSize);
        var dtos = comments.Select(c => new CommentAdminDto
        {
            Id = c.Id,
            UserId = c.UserId,
            Username = c.User?.Username ?? "deleted",
            Content = c.Content,
            TargetType = c.TargetType,
            TargetId = c.TargetId,
            CreatedAt = c.CreatedAt,
            DeletedAt = c.DeletedAt
        }).ToList();

        return new PaginatedResult<CommentAdminDto>
        {
            Items = dtos,
            Total = total,
            Page = page,
            PageSize = pageSize
        };
    }

    public async Task<bool> DeleteCommentAsync(int adminId, int commentId)
    {
        var success = await _commentRepo.DeleteAsync(commentId);
        if (success)
        {
            await _logRepo.LogAsync(new AdminLog
            {
                AdminId = adminId,
                Action = "delete_comment",
                TargetType = "comment",
                TargetId = commentId
            });
        }
        return success;
    }

    public async Task<bool> RestoreCommentAsync(int adminId, int commentId)
    {
        var success = await _commentRepo.RestoreAsync(commentId);
        if (success)
        {
            await _logRepo.LogAsync(new AdminLog
            {
                AdminId = adminId,
                Action = "restore_comment",
                TargetType = "comment",
                TargetId = commentId
            });
        }
        return success;
    }

    // --- Logs ---
    public async Task<PaginatedResult<AdminLogDto>> GetRecentLogsAsync(AdminLogFilterDto filter, int page = 1, int pageSize = 20)
    {
        var (logs, total) = await _logRepo.GetRecentAsync(filter, page, pageSize);
        var dtos = logs.Select(l => new AdminLogDto
        {
            Id = l.Id,
            AdminId = l.AdminId,
            AdminName = l.Admin?.Username ?? "unknown",
            Action = l.Action,
            TargetType = l.TargetType,
            TargetId = l.TargetId,
            Details = l.Details,
            CreatedAt = l.CreatedAt
        }).ToList();

        return new PaginatedResult<AdminLogDto>
        {
            Items = dtos,
            Total = total,
            Page = page,
            PageSize = pageSize
        };
    }
}

File: slp/backend-nestjs/src/modules/admin/admin.service.ts
import {
  Injectable,
  NotFoundException,
  BadRequestException,
} from "@nestjs/common";
import { UserRepository } from "../user/user.repository";
import { QuizRepository } from "../quiz/quiz.repository";
import { AdminLogRepository } from "./admin-log.repository";
import { UserDto } from "./dto/user.dto";
import { QuizAdminDto } from "./dto/quiz-admin.dto";
import { CommentAdminDto } from "./dto/comment-admin.dto";
import { AdminLogDto } from "./dto/admin-log.dto";
import { SessionRepository } from "../session/session.repository";
import { CommentRepository } from "../comment/comment.repository";

@Injectable()
export class AdminService {
  constructor(
    private readonly userRepo: UserRepository,
    private readonly quizRepo: QuizRepository,
    private readonly commentRepo: CommentRepository,
    private readonly sessionRepo: SessionRepository,
    private readonly logRepo: AdminLogRepository,
  ) {}

  // Users
  async getAllUsers(): Promise<UserDto[]> {
    const users = await this.userRepo.getAll(); // changed from findAll()
    return users.map((u) => ({
      id: u.id,
      username: u.username,
      email: u.email,
      emailConfirmed: u.emailConfirmed,
      role: u.role,
      status: u.status,
      createdAt: u.createdAt,
    }));
  }

  async getAllQuizzes(): Promise<QuizAdminDto[]> {
    const quizzes = await this.quizRepo.getAllForAdmin(); // changed from findAllWithUser()
    return quizzes.map((q) => ({
      id: q.id,
      title: q.title,
      userId: q.userId,
      username: q.user?.username || "deleted",
      visibility: q.visibility,
      disabled: q.disabled,
      createdAt: q.createdAt,
    }));
  }

  async disableQuiz(adminId: number, quizId: number): Promise<void> {
    const quiz = await this.quizRepo.getById(quizId, true); // changed from findById()
    if (!quiz) throw new NotFoundException("Quiz not found");

    quiz.disabled = true;
    await this.quizRepo.update(quiz);

    await this.logRepo.log(adminId, "disable_quiz", "quiz", quizId);
  }

  async enableQuiz(adminId: number, quizId: number): Promise<void> {
    const quiz = await this.quizRepo.getById(quizId, true);
    if (!quiz) throw new NotFoundException("Quiz not found");

    quiz.disabled = false;
    await this.quizRepo.update(quiz);

    await this.logRepo.log(adminId, "enable_quiz", "quiz", quizId);
  }

  async banUser(adminId: number, userId: number): Promise<void> {
    const user = await this.userRepo.getById(userId); // changed from findById()
    if (!user) throw new NotFoundException("User not found");
    if (user.role === "admin")
      throw new BadRequestException("Cannot ban an admin");

    user.status = "banned";
    await this.userRepo.update(user);

    // Revoke all sessions
    await this.sessionRepo.revokeAllForUser(userId);

    await this.logRepo.log(adminId, "ban_user", "user", userId);
  }

  async unbanUser(adminId: number, userId: number): Promise<void> {
    const user = await this.userRepo.getById(userId);
    if (!user) throw new NotFoundException("User not found");

    user.status = "active";
    await this.userRepo.update(user);

    await this.logRepo.log(adminId, "unban_user", "user", userId);
  }

  async getAllComments(
    includeDeleted: boolean = false,
  ): Promise<CommentAdminDto[]> {
    const comments = await this.commentRepo.findAllWithUser(includeDeleted);
    return comments.map((c) => ({
      id: c.id,
      userId: c.userId,
      username: c.user?.username || "deleted",
      content: c.content,
      targetType: c.targetType,
      targetId: c.targetId,
      createdAt: c.createdAt,
      deletedAt: c.deletedAt ?? undefined,
    }));
  }

  async deleteComment(adminId: number, commentId: number): Promise<void> {
    const success = await this.commentRepo.softDelete(commentId);
    if (!success) throw new NotFoundException("Comment not found");

    await this.logRepo.log(adminId, "delete_comment", "comment", commentId);
  }

  async restoreComment(adminId: number, commentId: number): Promise<void> {
    const success = await this.commentRepo.restore(commentId);
    if (!success) throw new NotFoundException("Comment not found");

    await this.logRepo.log(adminId, "restore_comment", "comment", commentId);
  }

  // Logs
  async getRecentLogs(limit: number = 100): Promise<AdminLogDto[]> {
    const logs = await this.logRepo.getRecent(limit);
    return logs.map((l) => ({
      id: l.id,
      adminId: l.adminId,
      adminName: l.admin?.username || "unknown",
      action: l.action,
      targetType: l.targetType ?? undefined,
      targetId: l.targetId ?? undefined,
      details: l.details,
      createdAt: l.createdAt,
    }));
  }
}
